src/buffer/out/textBuffer.cpp

```diff
@@ -1928,90 +1928,73 @@ void TextBuffer::_ExpandTextRow(til::inclusive_rect& textRow) const
-
-        const auto highlight = Viewport::FromInclusive(selectionRects.at(i));
-
-        // retrieve the data from the screen buffer
-        auto it = GetCellDataAt(highlight.Origin(), highlight);
-
-        // allocate a string buffer
         std::wstring selectionText;
-        std::vector<COLORREF> selectionFgAttr;
-        std::vector<COLORREF> selectionBkAttr;
+        til::small_rle<TextAttribute, uint16_t, 1> attrs;
 
-        // preallocate to avoid reallocs
-        selectionText.reserve(gsl::narrow<size_t>(highlight.Width()) + 2); // + 2 for \r\n if we munged it
-        if (copyTextColor)
-        {
-            selectionFgAttr.reserve(gsl::narrow<size_t>(highlight.Width()) + 2);
-            selectionBkAttr.reserve(gsl::narrow<size_t>(highlight.Width()) + 2);
-        }
+        const auto iRow = selectionRects.at(i).top;
+        const auto& row = GetRowByOffset(iRow);
+        const auto colBegin = gsl::narrow_cast<uint16_t>(selectionRects.at(i).left);
+        auto colEnd = gsl::narrow_cast<uint16_t>(selectionRects.at(i).right + 1); // end column exclusive
 
-        // copy char data into the string buffer, skipping trailing bytes
-        while (it)
+        // We apply formatting to rows if the row was NOT wrapped or formatting
+        // of wrapped rows is allowed.
+        const auto shouldFormatRow = formatWrappedRows || !row.WasWrapForced();
+
+        if (trimTrailingWhitespace && shouldFormatRow)
         {
-            const auto& cell = *it;
+            // update end column to exclude trailing whitespace
 
-            if (cell.DbcsAttr() != DbcsAttribute::Trailing)
+            const auto text = row.GetText();
+            const auto begin = text.rbegin();
+            const auto end = text.rend();
+            auto it = begin;
+            for (; it != end; ++it)
             {
-                const auto chars = cell.Chars();
-                selectionText.append(chars);
-
-                if (copyTextColor)
+                if (*it != L' ')
                 {
-                    const auto cellData = cell.TextAttr();
-                    const auto [CellFgAttr, CellBkAttr] = GetAttributeColors(cellData);
-                    for (size_t j = 0; j < chars.size(); ++j)
-                    {
-                        selectionFgAttr.push_back(CellFgAttr);
-                        selectionBkAttr.push_back(CellBkAttr);
-                    }
+                    break;
                 }
             }
 
-            ++it;
+            const auto textTrimmedLength = gsl::narrow_cast<uint16_t>(end - it);
+            colEnd = std::min(colEnd, textTrimmedLength);
         }
 
-        // We apply formatting to rows if the row was NOT wrapped or formatting of wrapped rows is allowed
-        const auto shouldFormatRow = formatWrappedRows || !GetRowByOffset(iRow).WasWrapForced();
+        // save selected text
+        selectionText = std::wstring{ row.GetText(colBegin, colEnd) };

```

Comment:
Side note, BTW: This text copying is among others a reason for why I really want to remove `TextAndAttribute` entirely. It'd be much more efficient but just as simple to accumulate the HTML/RTF data right here.

Reply:
If you mean to say accumulate HTML/RTF data for _this line_, I guess we can do this in these steps:
1. Generate HTML/RTF header data.
2. Generate HTML/RTF _attr-run-by-attr-run_ by calling `Gen[HTML|RTF]` in a loop. `Gen[HTML|RTF]` would then only be responsible for _this span of text_ (or an RTF group in case of RTF). We'll pass a single attribute which needs to be applied for this run of text.
3. Generate HTML/RTF footer data.
4. Final result would be Header + All text spans + Footer.

Does this make sense?

Reply:
What I meant is an approach like the one I linked above in this comment: https://github.com/microsoft/terminal/pull/16270#issuecomment-1797039935
It accumulates the text directly into `std::wstring& out`. (FYI it uses an out parameter because that's the only way the caller can control both, the console lock duration, and the batch size for `WriteFile` calls. For functions like `GenHTML` and `GenRTF`, I'd just return the `string`.)

I'd basically merge `GetText` into each `GenHTML` and `GenRTF` which then do their own text- and attribute-processing directly on the `ROW`s. Only once they're standalone functions without any external helpers, I'd start abstracting smaller bits away as needed.

Reply:
Ok.. I'm implementing this, just letting you know that we will be calling `RenderSettings::GetAttributeColors` twice now, each in `GenHTML` and `GenRTF`. Would this be a concern?

I'm also looking into having one external helper where we would be calculating the exact text span we're interested in from the selection region. This is what will be responsible for trimming trailing whitespaces. We do this once and pass the result into `GenHTML` and `GenRTF`.